### ISSUE TITLE
Safe GHC LANGUAGE extensions to be put into Agda.cabal

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#1942]
There is a couple of safe language extensions we keep repeating in each .hs file. I suggest to put all frequently-used and semantics preserving language extension into the Agda.cabal file instead.  These are extensions to the Haskell 98 syntax and safe extensions to the Haskell 98 type class systems.  I suggest the following ones:

```
BangPatterns
ConstraintKinds
DataKinds
DefaultSignatures
DeriveDataTypeable
DeriveFoldable
DeriveFunctor
DeriveGeneric
DeriveTraversable
DoAndIfThenElse
EmptyDataDecls
ExistentialQuantification
FlexibleContexts
FlexibleInstances
FunctionalDependencies
GADTs
GeneralizedNewtypeDeriving
LambdaCase
MultiParamTypeClasses
NamedFieldPuns
PatternGuards
RankNTypes
RecordWildCards
ScopedTypeVariables
StandaloneDeriving
TupleSections
TypeFamilies
TypeSynonymInstances
```

Purposely not in the list are

```
ImplicitParams
IncoherentInstances
MagicHash
NondecreasingIndentation
OverlappingInstances
RelaxedPolyRec
TemplateHaskell
UndecidableInstances
ViewPatterns
```

[Breakfast discussion with @UlfNorell]

**Comment by asr** [comment]
I disagree for various reasons. I start saying that default languages extensions makes it harder to understand the errors of compilation and to support three released versions of GHC.

To give a concrete example. I _just_ added the list of the extensions proposed and I got the error:

```
$ make install-bin
...
[ 46 of 334] Compiling Agda.Utils.Bag   ( src/full/Agda/Utils/Bag.hs, dist-2.5.1.1/build/Agda/Utils/Bag.o ) [flags changed]

src/full/Agda/Utils/Bag.hs:56:1:
    The type signature for ‘!’ lacks an accompanying binding
      (The type signature must be given where ‘!’ is declared)

src/full/Agda/Utils/Bag.hs:63:30:
    Not in scope: ‘!’
    Perhaps you meant ‘Map.!’ (imported from Data.Map)

src/full/Agda/Utils/Bag.hs:67:23:
    Not in scope: ‘!’
    Perhaps you meant ‘Map.!’ (imported from Data.Map)
```

Which extension generated the error?

**Comment by UlfNorell** [comment]
Stab in the dark: `BangPatterns`.

I would argue that having a default set of extensions make errors more easy to understand since you are using a (more) consistent language across modules. Also it really annoying to have to enable perfectly harmless language extensions.

**Comment by asr** [comment]
> Stab in the dark: `BangPatterns`.

You are right.

By removing `BangPatterns` I got the error:

```
$ make install-in
[ 73 of 334] Compiling Agda.Auto.NarrowingSearch ( src/full/Agda/Auto/NarrowingSearch.hs, dist-2.5.1.1/build/Agda/Auto/NarrowingSearch.o )

src/full/Agda/Auto/NarrowingSearch.hs:545:29:
    Couldn't match type ‘b1’ with ‘b2’
      ‘b1’ is a rigid type variable bound by
           a pattern with constructor
             PDoubleBlocked :: forall blk b1 b2.
                               (Refinable b1 blk, Refinable b2 blk) =>
                               Metavar b1 blk -> Metavar b2 blk -> MetaEnv (PB blk) -> PB blk,
           in a case alternative
           at src/full/Agda/Auto/NarrowingSearch.hs:541:5
      ‘b2’ is a rigid type variable bound by
           a pattern with constructor
             PDoubleBlocked :: forall blk b1 b2.
                               (Refinable b1 blk, Refinable b2 blk) =>
                               Metavar b1 blk -> Metavar b2 blk -> MetaEnv (PB blk) -> PB blk,
           in a case alternative
           at src/full/Agda/Auto/NarrowingSearch.hs:541:5
    Expected type: [(QPB b2 blk1, Maybe (CTree blk1))]
                   -> [(QPB b2 blk1, Maybe (CTree blk1))]
      Actual type: [(QPB b1 blk1, Maybe (CTree blk1))]
                   -> [(QPB b1 blk1, Maybe (CTree blk1))]
    Relevant bindings include
      newobs :: [(QPB b1 blk1, Maybe (CTree blk1))]
                -> [(QPB b1 blk1, Maybe (CTree blk1))]
        (bound at src/full/Agda/Auto/NarrowingSearch.hs:543:10)
      m2 :: Metavar b2 blk1
        (bound at src/full/Agda/Auto/NarrowingSearch.hs:541:23)
      m1 :: Metavar b1 blk1
        (bound at src/full/Agda/Auto/NarrowingSearch.hs:541:20)
    In the second argument of ‘umodifyIORef’, namely ‘newobs’
    In a stmt of a 'do' block: umodifyIORef (mobs m2) newobs
```

Which extension generated the new error?

**Comment by asr** [comment]
Other reason to disagree: Handle the default extensions is too Cabal-version dependent.

Examples:
- In recent versions of Cabal, the `extensions` field was removed and the `default-extensions` and `others-extensions` fields were added (see my comment [here](https://github.com/haskell/cabal/issues/1517#issuecomment-46843791)).
- From [here](https://github.com/haskell/cabal/pull/2818/commits/6916f3b53db19d4442914f4473e21669114025ba):
  
  In Cabal-1.24 the dependency solver will use this and `default-extensions` information. Cabal prior to 1.24 will abort compilation if the current compiler doesn't provide the extensions.

**Comment by asr** [comment]
(I updated the previous comment).

**Comment by UlfNorell** [comment]
> Which extension generated the new error?

This is a Wrong Question (although `ScopedTypeVariables`?). The question isn't which extension caused the error, but rather what's wrong with the code? It's like asking which module in GHC caused a type error.

> In recent versions of Cabal, the extensions field was removed...

If I'm reading that correctly, that's `cabal >= 1.10`. Given that a current version is `1.22`, surely we can require `>= 1.10`?

> In Cabal-1.24 the dependency solver will use this and default-extensions information. Cabal prior to 1.24 will abort compilation if the current compiler doesn't provide the extensions.

How is this situation improved by having the language pragmas in the modules?

**Comment by asr** [comment]
> If I'm reading that correctly, that's `cabal >= 1.10`. Given that a current version is `1.22`, surely we can require `>= 1.10`?

Yes. We are using `cabal-version: >= 1.8` for [this](https://github.com/judah/haskeline/issues/17#issuecomment-68946643) reason.

**Comment by asr** [comment]
> In Cabal-1.24 the dependency solver will use this and default-extensions information. Cabal prior to 1.24 will abort compilation if the current compiler doesn't provide the extensions.
> 
> > How is this situation improved by having the language pragmas in the modules?

I don't understand why the question is related to my comment. I was pointed out an example where different versions of Cabal have different behaviour related to the default extensions.

**Comment by UlfNorell** [comment]
> Yes. We are using cabal-version: >= 1.8 for this reason.

I don't understand the reason. It looks like that issue suggests you should use >= 1.10?

> I don't understand why the question is related to my comment. I was pointed an example where different versions of Cabal have different behaviour related to the default extensions.

Yes, but it looks to me like newer versions of cabal will behave better if you put extensions in the cabal file. I don't see a reason to cripple later cabal versions just to make them behave the same as older ones.

**Comment by asr** [comment]
> It looks like that issue suggests you should use >= 1.10?

I understood the opposite, that is, it's not necessary to use >= 1.10.

**Comment by UlfNorell** [comment]
> I understood the opposite, that is, it's not necessary to use >= 1.10.

It is if we want to use the `default-extensions` field. Is there a reason we cannot require `>= 1.10`?

**Comment by asr** [comment]
> Yes, but it looks to me like newer versions of cabal will behave better if you put extensions in the cabal file.

I agree.

> I don't see a reason to cripple later cabal versions just to make them behave the same as older ones.

For supporting the old ones (I guess some distributions don't use the latest version of Cabal). Our current  restriction (i.e. `cabal-version: >= 1.8`) is not a real one and we test on Travis the different versions of GHC only using the latest version of Cabal.

**Comment by asr** [comment]
> Is there a reason we cannot require >= 1.10?

I don't know any reason.

**Comment by UlfNorell** [comment]
> For supporting the old ones (I guess some distributions don't use the latest version of Cabal).

Yes, but in this case older versions of cabal behave exactly the same regardless of whether you put the extensions in the cabal file or in the Haskell modules, so I fail to see how that argument applies to the current issue.

**Comment by asr** [comment]
> Yes, but in this case older versions of cabal behave exactly the same regardless of whether you put the extensions in the cabal file or in the Haskell modules

No, if you exclude those older versions of Cabal.

**Comment by asr** [comment]
Instead of adding the "deriving extensions"  we should remove them. This is a common advice for reducing the compilation time.

**Comment by asr** [comment]
I would like to compile Agda using SAFE Haskell. The last time I checked, we couldn't use `GeneralizedNewtypeDeriving` nor `TemplateHaskell`.

**Comment by UlfNorell** [comment]
> No, if you exclude those older versions of cabal.

Are you saying we need to support distributions running `cabal 1.8`? For reference, that was released around the same time as `ghc-6.12`.

> Instead of adding the "deriving extensions" but should remove them. This is a common advice for reducing the compilation time.

Do you have any sources for this? I have a very hard time believing compilation times goes up just because you enable `DeriveFunctor` (say) without actually deriving any Functor instances. If they do that sounds like a pretty bad bug in GHC.

> I would like to compile Agda using SAFE Haskell. The last time I checked, we couldn't use GeneralizedNewtypeDeriving nor TemplateHaskell.

This issue isn't about which extensions we enable in the code, it's about where we write them. Also `TemplateHaskell` was on the list of extensions that should not go in the cabal file.

**Comment by asr** [comment]
> This is a Wrong Question (although ScopedTypeVariables?)

Nope.

**Comment by asr** [comment]
> Are you saying we need to support distributions running cabal 1.8? For reference, that was released around the same time as ghc-6.12.

No. I was pointing out that currently we don't exclude any recent version of Cabal (I don't remember any issue related to Cabal). If we add the default extensions then to keep the support for different recent versions of Cabal could be more difficult.

**Comment by asr** [comment]
> Do you have any sources for this? I have a very hard time believing compilation times goes up just because you enable DeriveFunctor (say) without actually deriving any Functor instances. If they do that sounds like a pretty bad bug in GHC.

I didn't say that. What I mean is that we should remove the _current_  "deriving extensions"  for reducing compilation times. By adding these extensions as default ones we are going in the opposite direction.

**Comment by asr** [comment]
I got the error reported [here](https://github.com/agda/agda/issues/1942#issuecomment-213065349) using GHC 7.10.3. By using GHC 8.0.1-rc3 I got a different error (I know it is a release candidate).

**Comment by UlfNorell** [comment]
> No. I was pointing out that currently we don't exclude any recent version of Cabal (I don't remember any issue related to Cabal). If we add the default extensions then to keep the support for different recent versions of Cabal could be more difficult.

I don't understand what you mean here. All recent versions of cabal support the `default-extensions` flag.

> I didn't say that. What I mean is that we should remove the current "deriving extensions" for reducing compilation times. By adding these extensions as default ones we are going in the opposite direction.

That's insane. Writing instances by hand is error prone and tedious and leads to extra work every time a data type changes. You would need some extremely solid evidence that there's a performance problem with deriving instances before even considering getting rid of those extensions. 

> I got the error reported here using GHC 7.10.3. By using GHC 8.0.1-rc3 I got a different error (I know it is a release candidate).

A better question is if the correct code works with both (all) versions of ghc. That they've changed an error message in GHC 8 is not that surprising.

**Comment by asr** [comment]
> A better question is if the correct code works with both (all) versions of ghc.

I agree.

> That they've changed an error message in GHC 8 is not that surprising.

(Since I agree with your previous comment just for clarifying) I'm not talking about a different error message, I got an error in a different module.

**Comment by asr** [comment]
> I don't understand what you mean here. All recent versions of cabal support the `default-extensions` flag.

I want to avoid something like `cabal-version: >= A.B.C` where A.B.C is a very recent version of Cabal. I know that this is currently not the case.

**Comment by asr** [comment]
> That's insane. Writing instances by hand is error prone and tedious and leads to extra work every time a data type changes. You would need some extremely solid evidence that there's a performance problem with deriving instances before even considering getting rid of those extensions.

See section "Fast builds lead to faster turnaround" on  [The Joy and Agony of Haskell in Production](http://www.stephendiehl.com/posts/production.html).

**Comment by asr** [comment]
Are the Cabal default options included in the modules generated by Alex and Happy?

**Comment by UlfNorell** [comment]
> See section "Fast builds lead to faster turnaround" on The Joy and Agony of Haskell in Production.

Quoting from that:

> The somewhat ugly hack to speed up compile time here is to run ghc -ddump-deriv Module.hs and then manually insert the resulting code instead of deriving it everytime.

Reading this as "you shouldn't using deriving" is clearly insane. It's saying that if you have a problem with deriving taking a long time, here's an ugly workaround. Are you saying we have a big problem with compilation times for derived instances?

> Are the Cabal default options included in the modules generated by Alex and Happy?

Yes of course.

**Comment by asr** [comment]
> Are you saying we have a big problem with compilation times for derived instances?

No. I'm saying I'm not comfortable with compilation times. Are the derived instances one important factor in the current compilation times?  I don't know but it's worth finding out.

**Comment by UlfNorell** [comment]
Sure, but it's not really relevant to this issue.

**Comment by asr** [comment]
Note the CPP was not mentioned in the OP.

**Comment by andreasabel** [comment]
I suggest to include all safe language pragmas in the cabal file that are sufficiently popular, e.g. are used in at least 5 of our Agda source files (this threshold is negotiable, could be 10).

**Comment by asr** [comment]
I know I wont' win the discussion :-)

Anyway, I'm strongly against of including `GeneralizedNewtypeDeriving`.

Some suggests:
- Right now, Agda supports GHC 8.0.1-rc4 (except the UHC backend). Before adding the defaults extensions I suggest to add a Travis instance for this version of GHC (which is possible in our current Travis set up).
- I suggest to keep the current local extensions for debugging purposes.

**Comment by Blaisorblade** [comment]
I'm taking a look at this. The type error in https://github.com/agda/agda/issues/1942#issuecomment-213065349 is introducing by adding at least one of GADTs or TypeFamilies.

I'm trying to add all extensions in the list, minus those debates or that introduce at least one compile error.

The experiment shows that adding a "safe" extension can still introduce compile errors — but we can detect those easily. Silent changes of meaning would be much worse, but I guess it isn't possible?

EIther way, I'll proceed with care.

**Comment by Blaisorblade** [comment]
I've noticed that NondecreasingIndentation was excluded on purpose. However, it's enabled by default _and_ disabled by Haskell2010, and it's used in at least one file:

```
    [226 of 338] Compiling Agda.TypeChecking.Reduce ( src/full/Agda/TypeChecking/Reduce.hs, .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/Agda/TypeChecking/Reduce.o ) [flags changed]

    /Users/pgiarrusso/AeroFS/Repos/agda-tatooine/src/full/Agda/TypeChecking/Reduce.hs:516:70: Empty 'do' block

    /Users/pgiarrusso/AeroFS/Repos/agda-tatooine/src/full/Agda/TypeChecking/Reduce.hs:525:13: Empty 'do' block

    /Users/pgiarrusso/AeroFS/Repos/agda-tatooine/src/full/Agda/TypeChecking/Reduce.hs:536:91: Empty 'do' block

    /Users/pgiarrusso/AeroFS/Repos/agda-tatooine/src/full/Agda/TypeChecking/Reduce.hs:576:100: Empty 'do' block

    /Users/pgiarrusso/AeroFS/Repos/agda-tatooine/src/full/Agda/TypeChecking/Reduce.hs:682:49: Empty 'do' block
```

**Comment by asr** [comment]
> I suggest to keep the current local extensions for debugging purposes.

I was surprised that so many language extensions were accepted by the four versions of GHC, so I changed my mind on the above suggestion.

**Comment by Blaisorblade** [comment]
@asr Local extensions = the language directives in sources, right? Removing them sounds good, but doing it eagerly might be lots of work. I'll see if I can get a decent result semi-automatically with `grep -v`, but I'll need to review the result.

**Comment by asr** [comment]
Right.

**Comment by Blaisorblade** [comment]
Also: what I enabled were just the extensions that could be enabled with minimal changes. I'm sure one could try adding the other ones that caused problems (ScopedTypeVariables, GADTs, BangPatterns, TypeFamilies), one would just have to do it one extension at a time. But I'd say that's for after #2017.

**Comment by asr** [comment]
The issue was closed by GitHub after #2030 was merged. I'm reopening it.

**Comment by Blaisorblade** [comment]
Now that #2017 is merged I'm tempted to close this for now, since OP @andreasabel agreed with the list of extensions.

Things like GADTs & c. should probably also be added, but separately.

**Comment by andreasabel** [comment]
Thanks, Paolo!

In the end, the `deriving` extensions did not make it into the patch, like

```
GeneralizedNewtypeDeriving
DeriveDataFunctor
...
```

I am not fully clear why they are excluded.  We use them quite frequently.  Does anything break if we turn them on globally?

**Comment by Blaisorblade** [comment]
Nothing should break but I haven't tried, that's just because the last
specific comment on them was an objection, and I've started from
uncontroversial stuff. But I don't buy the objection, people aren't
omitting instances just because they need extensions for them.
However, GeneralizedNewtypeDeriving was not type safe (because of
interaction with GADTs), until they added roles in GHC 7.8 simply to fix
the issue (forgot the reason). But Agda still supports GHC 7.6. (OTOH, all
uses are accepted under later GHCs, so arguably it's safe).
